src/metabase/query_processor/parameters.clj

```diff
@@ -10,18 +10,116 @@
             [metabase.util :as u])
   (:import (org.joda.time DateTimeConstants DateTime)))
 
+(defn- day-range
+  [^DateTime end ^DateTime start]
+  {:end   end
+   :start start})
 
-(def ^:private ^:const relative-dates
-  #{"today"
-    "yesterday"
-    "past7days"
-    "past30days"
-    "thisweek"
-    "thismonth"
-    "thisyear"
-    "lastweek"
-    "lastmonth"
-    "lastyear"})
+(defn- week-range
+  [^DateTime end ^DateTime start]
+    ;; weeks always start on SUNDAY and end on SATURDAY
+    ;; NOTE: in Joda the week starts on Monday and ends on Sunday, so to get the right Sunday we rollback 1 week
+   {:end   (.withDayOfWeek end DateTimeConstants/SATURDAY)
+    :start (.withDayOfWeek ^DateTime (t/minus start (t/weeks 1)) DateTimeConstants/SUNDAY)})
+
+(defn- month-range
+  [^DateTime end ^DateTime start]
+  {:end   (t/last-day-of-the-month end)
+   :start (t/first-day-of-the-month start)})
+
+(defn- year-range
+  [^DateTime end ^DateTime start]
+  {:end   (t/last-day-of-the-month  (.withMonthOfYear end DateTimeConstants/DECEMBER))
+   :start (t/first-day-of-the-month (.withMonthOfYear start DateTimeConstants/JANUARY))})
+
+(def ^:private operations-by-date-unit
+  {"day"   {:unit-range     day-range
+            :to-period t/days}
+   "week"  {:unit-range     week-range
+            :to-period t/weeks}
+   "month" {:unit-range     month-range
+            :to-period t/months}
+   "year"  {:unit-range     year-range
+            :to-period t/years}})
+
+(defn ^:private parse-absolute-date
+  [date]
+  (tf/parse (tf/formatters :date-parser) date))
+
+(defn ^:private expand-parser-groups
+  [group-label group-value]
+  (case group-label
+    :unit (conj (seq (get operations-by-date-unit group-value))
+                [group-label group-value])
+    :int-value [[group-label (Integer/parseInt group-value)]]
+    (:date :date-1 :date-2) [[group-label (parse-absolute-date group-value)]]
+    [[group-label group-value]]))
+
+(defn ^:private regex->parser
+  "Takes a regex and labels matching the regex capturing groups. Returns a parser which
+  takes a parameter value, validates the value against regex and gives a map of labels
+  and group values. Respects the following special label names:
+      :unit – finds a matching date unit and merges date unit operations to the result
+      :int-value – converts the group value to integer
+      :date[-index] – converts the group value to absolute date"
+  [regex, group-labels]
+  (fn [param-value]
+    (when-let [regex-result (re-matches regex param-value)]
+      (into {}
+             (mapcat expand-parser-groups group-labels (rest regex-result))))))
```

Comment:
Args should be aligned if they're on newlines, e.g.

```clojure
(into {}
      (...))
```

(looks like you're off by a space)

But FWIW I don't think putting this on a newline is really needed here